src/validators/AllowedSessionsValidator.sol

```diff
@@ -0,0 +1,127 @@
+import { IValidatorManager } from "../interfaces/IValidatorManager.sol";
+import { SessionLib } from "../libraries/SessionLib.sol";
+import { Errors } from "../libraries/Errors.sol";
+import { SignatureDecoder } from "../libraries/SignatureDecoder.sol";
+
+import { SessionKeyValidator } from "./SessionKeyValidator.sol";
+
+/// @title AllowedSessionsValidator
+/// @author Oleg Bedrin - <o.bedrin@xsolla.com> - Xsolla Web3
+/// @custom:security-contact security@matterlabs.dev and o.bedrin@xsolla.com
+/// @notice This contract is used to manage allowed sessions for a smart account.
+contract AllowedSessionsValidator is SessionKeyValidator, AccessControl {
+  using SessionLib for SessionLib.SessionStorage;
+
+  /// @notice Role identifier for session registry managers.
+  bytes32 public constant SESSION_REGISTRY_MANAGER_ROLE = keccak256("SESSION_REGISTRY_MANAGER_ROLE");
+
+  /// @notice Mapping to track whether a session actions is allowed.
+  /// @dev The key is the hash of session actions, and the value indicates if the actions are allowed.
+  mapping(bytes32 sessionActionsHash => bool active) public isSessionSpecAllowed;
+
+  /// @notice Emitted when session actions are allowed or disallowed.
+  /// @param sessionActionsHash The hash of the session actions.
+  /// @param allowed Boolean indicating if the session actions are allowed.
+  event SessionActionsAllowed(bytes32 sessionActionsHash, bool allowed);
+
+  /// @notice Error indicating that the session actions are not allowed.
+  /// @param sessionActionsHash The hash of the session actions that are not allowed.
+  error SessionActionsNotAllowed(bytes32 sessionActionsHash);
+
+  constructor() {
+    _grantRole(SESSION_REGISTRY_MANAGER_ROLE, msg.sender);
+    _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+  }
+
+  /// @notice Set whether a session actions hash is allowed or not.
+  /// @param sessionActionsHash The hash of the session actions.
+  /// @param allowed Boolean indicating if the session actions are allowed.
+  /// @dev Session actions represent the set of operations, such as fee limits, call policies, and transfer policies,
+  /// that define the behavior and constraints of a session.
+  function setSessionActionsAllowed(
+    bytes32 sessionActionsHash,
+    bool allowed
+  ) external virtual onlyRole(SESSION_REGISTRY_MANAGER_ROLE) {
+    isSessionSpecAllowed[sessionActionsHash] = allowed;
+    emit SessionActionsAllowed(sessionActionsHash, allowed);
+  }
+
+  /// @notice Get the hash of session actions from a session specification.
+  /// @param sessionSpec The session specification.
+  /// @return The hash of the session actions.
+  /// @dev The session actions hash is derived from the session's fee limits, call policies, and transfer policies.
+  function getSessionActionsHash(SessionLib.SessionSpec memory sessionSpec) public view virtual returns (bytes32) {
+    return keccak256(abi.encode(sessionSpec.feeLimit, sessionSpec.callPolicies, sessionSpec.transferPolicies));
```

Comment:
I think including full `callPolicies` in this hash goes against what we discussed previously on a call. Since `callPolicies` include the `Constraints[]` array, which is very likely to include account-specific constraints (e.g. argument constrained to account's own address). Which would make such constraints impossible to pass unless whitelisted separately for each user.

Reply:
Hm, that's right, I forgot that! I'll figure something out and re-ping you here or in the chat. Thanks!